myokit/formats/diffsl/_ewriter.py

```diff
@@ -0,0 +1,205 @@
+        warnings.warn('Unsupported function: acos()')
+        return super()._ex_acos(e)
+
+    def _ex_asin(self, e):
+        warnings.warn('Unsupported function: asin()')
+        return super()._ex_asin(e)
+
+    def _ex_atan(self, e):
+        warnings.warn('Unsupported function: atan()')
+        return super()._ex_atan(e)
+
+    def _ex_ceil(self, e):
+        warnings.warn('Unsupported function: ceil()')
+        return super()._ex_ceil(e)
+
+    # def _ex_cos(self, e):
+    # def _ex_derivative(self, e):
+    # def _ex_divide(self, e):
+    # def _ex_exp(self, e):
+
+    def _ex_floor(self, e):
+        warnings.warn('Unsupported function: floor()')
+        return super()._ex_floor(e)
+
+    # def _ex_log(self, e):
+
+    def _ex_log10(self, e):
+        # Log10(a) = Log(a, 10.0) -> '(log(a) / log(10.0))'
+        return super()._ex_log(Log(e[0], Number(10)))
+
+    # def _ex_minus(self, e):
+    # def _ex_multiply(self, e):
+    # def _ex_plus(self, e):
+    # def _ex_power(self, e):
+    # def _ex_prefix_minus(self, e):
+    # def _ex_prefix_plus(self, e):
+    # def _ex_quotient(self, e):
+    # def _ex_remainder(self, e):
+    # def _ex_sin(self, e):
+    # def _ex_sqrt(self, e):
+    # def _ex_tan(self, e):
+
+    # -- Conditional operators
+
+    def _ex_and(self, e):
+        # (a and b) == a * b, where a, b are in {0, 1}
+        return f'{self.ex(e[0])} * {self.ex(e[1])}'
+
+    def _ex_equal(self, e):
+        # (a == b) == heaviside(a - b) * heaviside(b - a)
+        return self.ex(And(MoreEqual(e[0], e[1]), LessEqual(e[0], e[1])))
+
+    def _ex_less(self, e):
+        # (a < b) == 1 - heaviside(a - b)
+        return self.ex(Not(MoreEqual(e[0], e[1])))
+
+    def _ex_less_equal(self, e):
+        # (a <= b) == heaviside(b - a)
+        return f'heaviside({self.ex(e[1])} - {self.ex(e[0])})'
+
+    def _ex_more(self, e):
+        # (a > b) == 1 - heaviside(b - a)
+        return self.ex(Not(LessEqual(e[0], e[1])))
+
+    def _ex_more_equal(self, e):
+        # (a >= b) == heaviside(a - b)
+        return f'heaviside({self.ex(e[0])} - {self.ex(e[1])})'
+
+    def _ex_not(self, e):
+        # Simplify expressions with double not's.
+        # Examples:
+        # not(not(a)) -> (1 - (1 - a)) == a
+        # not(a or b) -> (1 - (1 - (1 - a) * (1 - b))) == (1 - a) * (1 - b)
+        if isinstance(e[0], Not):
+            # not(not(a)) == a
+            return self.ex(e[0][0])
+
+        if isinstance(e[0], NotEqual):
+            # not(not(a == b)) == (a == b)
+            return self.ex(Equal(e[0][0], e[0][1]))
+
+        if isinstance(e[0], Less):
+            # not(a < b) == (a >= b)
+            return self.ex(MoreEqual(e[0][0], e[0][1]))
+
+        if isinstance(e[0], More):
+            # not(a > b) == (a <= b)
+            return self.ex(LessEqual(e[0][0], e[0][1]))
+
+        if isinstance(e[0], Or):
+            # not(a or b) == not(a) and not(b)
+            return self.ex(And(Not(e[0][0]), Not(e[0][1])))
```

Comment:
With a few exceptions, I'm not sure this is a good idea!

1. Changing things at this point can lead to tricky bugs if the original and replaced expressions have different operator precedence. For this reason, this kind of optimisation should be made at the symbolic stage, not in the expression writer.
2. I'm not 100% convinced we need these. If the user writes a silly model it's ok for us to generate silly code.

Reply:
This has been resolved by 7ac6407f